lib/SILOptimizer/IPO/CapturePromotion.cpp

```diff
@@ -350,18 +350,21 @@ computeNewArgInterfaceTypes(SILFunction *F,
     assert(paramBoxTy->getLayout()->getFields().size() == 1
            && "promoting compound box not implemented yet");
     auto paramBoxedTy = paramBoxTy->getFieldType(F->getModule(), 0);
-    auto &paramTL = F->getModule().Types.getTypeLowering(paramBoxedTy);
-
-    ParameterConvention convention;
-    if (paramTL.isFormallyPassedIndirectly()) {
-      convention = ParameterConvention::Indirect_In;
-    } else if (paramTL.isTrivial()) {
-      convention = ParameterConvention::Direct_Unowned;
-    } else {
-      convention = ParameterConvention::Direct_Owned;
+    {
+      Lowering::GenericContextScope GenericScope(
+          F->getModule().Types, paramBoxTy->getLayout()->getGenericSignature());
+      auto &paramTL = F->getTypeLowering(paramBoxedTy);
```

Comment:
@jckarter I think I found the actual issue. The current code in the master uses:
``` 
 auto &paramTL = F->getModule().Types.getTypeLowering(paramBoxedTy);
```
whereas the right thing to do would be:
``` 
 auto &paramTL = F->getTypeLowering(paramBoxedTy);
```

`SILFunction:: getTypeLowering` actually pushes the right generic context, etc.

I'll try to come up with a reduced test-case.